proc/process.c

```diff
@@ -395,26 +395,186 @@ static void process_tlsAssign(hal_tls_t *process_tls, hal_tls_t *tls, ptr_t tbss
+			return -ENOEXEC;
+		}
+	}
+
+	return 0;
+}
+
+
 #ifndef NOMMU
 
+
+/* There are no NOMMU 64 bit targets.. */
+static int process_loadValidateElf64(void *iehdr, size_t size)
+{
+	Elf64_Ehdr *ehdr = iehdr;
+	Elf64_Phdr *phdr;
+	Elf64_Shdr *shdr, *shstrshdr;
+	char *snameTab;
+	size_t memsz, filesz;
+	unsigned i, misalign;
+	offs_t offs;
+
+	if (process_validatePtr(iehdr, size, ehdr, sizeof(*ehdr)) == 0) {
+		return -ENOEXEC;
+	}
+
+	/* Validate header. */
+	if (((process_validatePtr(iehdr, size, ehdr->e_ident, 4) == 0) ||
+			(hal_strncmp((char *)ehdr->e_ident, "\177ELF", 4) == 0)) ||
+		(ehdr->e_shnum == 0)) {
+		return -ENOEXEC;
+	}
+
+	phdr = (void *)ehdr + ehdr->e_phoff;
+	if (process_validatePtr(iehdr, size, phdr, sizeof(*phdr) * ehdr->e_phnum) == 0) {
+		return -ENOEXEC;
+	}
+	for (i = 0; i < ehdr->e_phnum; i++) {
+		if ((phdr->p_type != PT_LOAD) &&
+			(process_validatePtr(iehdr, size, ((char *)ehdr) + phdr[i].p_offset, phdr[i].p_filesz) == 0)) {
+			return -ENOEXEC;
+		}
+
+		offs = phdr->p_offset & ~(phdr->p_align - 1);
+		misalign = phdr->p_offset & (phdr->p_align - 1);
+		filesz = phdr->p_filesz ? (phdr->p_filesz + misalign) : 0;
+		memsz = phdr->p_memsz + misalign;
+		if ((offs >= size) || (memsz < filesz)) {
+			return -ENOEXEC;
+		}
+	}
+
+	shdr = (void *)((char *)ehdr + ehdr->e_shoff);
+	if (process_validatePtr(iehdr, size, shdr, sizeof(*shdr) * ehdr->e_shnum) == 0) {
+		return -ENOEXEC;
+	}
+
+	shstrshdr = shdr + ehdr->e_shstrndx;
+	if (process_validatePtr(iehdr, size, shstrshdr, sizeof(*shstrshdr)) == 0) {
+		return -ENOEXEC;
+	}
+	snameTab = (char *)ehdr + shstrshdr->sh_offset;
+	if (process_validatePtr(iehdr, size, snameTab, shstrshdr->sh_size) == 0) {
+		return -ENOEXEC;
+	}
+	/* Strings must end with NULL character. */
+	if (snameTab[shstrshdr->sh_size - 1] != '\0') {
+		return -ENOEXEC;
+	}
+
+	for (i = 0; i < ehdr->e_shnum; i++) {
+		if ((process_validatePtr(iehdr, size, ((char *)ehdr) + shdr[i].sh_offset, shdr[i].sh_size) == 0) ||
+			(shdr->sh_name >= shstrshdr->sh_size)) {
+			return -ENOEXEC;
+		}
+	}
+
+	return 0;
+}
+
+
 /* TODO - adding error handling and unmapping of already mapped segments */
-int process_load32(vm_map_t *map, vm_object_t *o, offs_t base, void *iehdr, size_t *ustacksz, hal_tls_t *tls, ptr_t *tbssAddr)
+int process_load32(vm_map_t *map, vm_object_t *o, offs_t base, void *iehdr, size_t size, size_t *ustacksz, hal_tls_t *tls, ptr_t *tbssAddr)
 {
-	void *vaddr = NULL;
-	size_t memsz = 0, filesz = 0;
+	void *vaddr;
+	size_t memsz, filesz;
 	Elf32_Ehdr *ehdr = iehdr;
 	Elf32_Phdr *phdr;
-	Elf32_Shdr *shdr;
-	unsigned i, prot, flags, misalign = 0;
+	Elf32_Shdr *shdr, *shstrshdr;
+	unsigned i, prot, flags, misalign;
+	int ret;
 	offs_t offs;
 	char *snameTab;
 
-	shdr = (void *)((char *)ehdr + ehdr->e_shoff);
-	shdr += ehdr->e_shstrndx;
-	snameTab = (char *)ehdr + shdr->sh_offset;
+	ret = process_loadValidateElf32(iehdr, size);
+	if (ret < 0) {
+		return ret;
+	}
```

Comment:
Imho we don't really need retval of process_loadValidateElf32, as it either return -ENOEXEC or 0
```suggestion
	if (process_loadValidateElf32(iehdr, size) != 0) {
		return -ENOEXEC;
	}
```